src/shared/utils/config-utils/get-config.ts

```diff
@@ -0,0 +1,19 @@
+import { ConfigElements } from "../../constants";
+import { ConfigCache } from "../../types";
+import { getConfigFile } from "./s3-config-client";
+
+export type CachedConfigPromisesByElement = Partial<{
+  [T in ConfigElements]: Promise<ConfigCache[T]>;
+}>;
+
+const cachedConfigPromisesByFile: CachedConfigPromisesByElement = {};
+
+export const getConfig = async <T extends ConfigElements>(
+  fileName: T
+): Promise<ConfigCache[T]> => {
+  const cachedPromise = cachedConfigPromisesByFile[fileName];
+  if (cachedPromise !== undefined) return await cachedPromise;
+  const promise = getConfigFile(fileName);
+  (cachedConfigPromisesByFile[fileName] as Promise<ConfigCache[T]>) = promise;
+  return await promise;
+};
```

Comment:
Found a way to avoid coercion :D cache the files not the promises

```suggestion
export type CachedConfigByElement = Partial<{
  [T in ConfigElements]: ConfigCache[T];
}>;

const cachedConfigByFile: CachedConfigByElement = {};

export const getConfig = async <T extends ConfigElements>(
  fileName: T
): Promise<ConfigCache[T]> => {
  const cachedFile = cachedConfigByFile[fileName];
  if (cachedFile !== undefined) return cachedFile;
  const configFile = await getConfigFile(fileName);
  cachedConfigByFile[fileName] = configFile;
  return configFile;
};
```

Reply:
ok not this, this'll make a new promise if one is already in flight